### PR TITLE
Add DB alias logger middleware

### PR DESCRIPTION
**Pull request by maarcingebala**
Changes addressing RFC https://github.com/saleor/saleor/discussions/15407.

This PR adds a middleware that logs SQL queries with a DB alias - the name of the database that is used with each query. The purpose is to avoid calls using the `default` alias but explicitly pass either `replica` or `writer` with the `queryset.using()` function. For that purpose a new DB alias was introduced:

```
DATABASE_CONNECTION_WRITER_NAME = "writer"
```

### Configuration

To enable, set the env variable to True: `ENABLE_DB_ALIAS_LOGGER_MIDDLEWARE`.

The logger has the following configuration options:

```
DB_ALIAS_LOGGER = {
    "LOG_REPLICA": False,
    "LOG_WRITER": False,
}
```

By default, the replica and writer queries are not logged; only queries to the `default` alias are logged. To include replica or writer queries, switch the `LOG_REPLICA` and `LOG_WRITER` configuration options in settings. 

### Examples

Logging replica queries:
```
2024-02-20 08:35:41,331 INFO saleor.core.db.connection [db:replica]: SELECT t.oid, typarray FROM pg_type t JOIN pg_namespace ns ON typnamespace = ns.oid WHERE typname = 'hstore' [PID:77484:ThreadPoolExecutor-0_0]
2024-02-20 08:35:41,337 INFO saleor.core.db.connection [db:replica]: SELECT typarray FROM pg_type WHERE typname = 'citext' [PID:77484:ThreadPoolExecutor-0_0]
2024-02-20 08:35:41,338 INFO saleor.core.db.connection [db:replica]: SELECT "account_user"."id", "account_user"."is_superuser", "account_user"."private_metadata", "account_user"."metadata", "account_user"."external_reference", "account_user"."password", "account_user"."last_login", "account_user"."email", "account_user"."first_name", "account_user"."last_name", "account_user"."is_staff", "account_user"."is_active", "account_user"."is_confirmed", "account_user"."last_confirm_email_request", "account_user"."note", "account_user"."date_joined", "account_user"."updated_at", "account_user"."last_password_reset_request", "account_user"."default_shipping_address_id", "account_user"."default_billing_address_id", "account_user"."avatar", "account_user"."jwt_token_key", "account_user"."language_code", "account_user"."search_document", "account_user"."uuid" FROM "account_user" WHERE ("account_user"."is_active" AND "account_user"."email" IN (%s)) [PID:77484:ThreadPoolExecutor-0_0]
2024-02-20 08:35:41,349 INFO saleor.core.db.connection [db:replica]: SELECT "account_address"."id", "account_address"."private_metadata", "account_address"."metadata", "account_address"."first_name", "account_address"."last_name", "account_address"."company_name", "account_address"."street_address_1", "account_address"."street_address_2", "account_address"."city", "account_address"."city_area", "account_address"."postal_code", "account_address"."country", "account_address"."country_area", "account_address"."phone" FROM "account_address" INNER JOIN "account_user_addresses" ON ("account_address"."id" = "account_user_addresses"."address_id") WHERE ("account_user_addresses"."user_id" = %s AND "account_address"."id" = %s) ORDER BY "account_address"."id" ASC LIMIT 1 [PID:77484:ThreadPoolExecutor-0_0]
```

Logging writer queries:
```
2024-02-20 08:52:18,650 INFO saleor.core.db.connection [db:writer]: SELECT t.oid, typarray FROM pg_type t JOIN pg_namespace ns ON typnamespace = ns.oid WHERE typname = 'hstore' [PID:83044:ThreadPoolExecutor-0_0]
2024-02-20 08:52:18,665 INFO saleor.core.db.connection [db:writer]: SELECT typarray FROM pg_type WHERE typname = 'citext' [PID:83044:ThreadPoolExecutor-0_0]
2024-02-20 08:52:18,667 INFO saleor.core.db.connection [db:writer]: SELECT "account_user"."id", "account_user"."is_superuser", "account_user"."private_metadata", "account_user"."metadata", "account_user"."external_reference", "account_user"."password", "account_user"."last_login", "account_user"."email", "account_user"."first_name", "account_user"."last_name", "account_user"."is_staff", "account_user"."is_active", "account_user"."is_confirmed", "account_user"."last_confirm_email_request", "account_user"."note", "account_user"."date_joined", "account_user"."updated_at", "account_user"."last_password_reset_request", "account_user"."default_shipping_address_id", "account_user"."default_billing_address_id", "account_user"."avatar", "account_user"."jwt_token_key", "account_user"."language_code", "account_user"."search_document", "account_user"."uuid" FROM "account_user" WHERE ("account_user"."is_active" AND "account_user"."email" IN (%s)) [PID:83044:ThreadPoolExecutor-0_0]
```

# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/settings.py [hunk 1]
```diff
@@ -99,11 +99,20 @@ def get_url_from_env(name, *, schemes=None) -> Optional[str]:
...
+
 DATABASES = {
     DATABASE_CONNECTION_DEFAULT_NAME: dj_database_url.config(
         default="postgres://saleor:saleor@localhost:5432/saleor",
         conn_max_age=DB_CONN_MAX_AGE,
     ),
+    DATABASE_CONNECTION_WRITER_NAME: dj_database_url.config(
+        default="postgres://saleor:saleor@localhost:5432/saleor",
+        conn_max_age=DB_CONN_MAX_AGE,
+    ),
```
Comment (fowczarek) [inline saleor/settings.py:115]: If we release it create an additional writer connection for each instance.

**Comment by IKarbowiak** [comment]
IMO, we should document it somehow.

**Review by IKarbowiak — APPROVED** [review]

File: saleor/core/db/connection.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+def wrap_connection(connection: BaseDatabaseWrapper):
+    if not hasattr(connection, "_orig_cursor"):
+        connection._orig_cursor = connection.cursor  # type: ignore
+
+        def cursor(*args, **kwargs):
+            orig_cursor = connection._orig_cursor(*args, **kwargs)  # type: ignore
+            PatchedCursor = _apply_mixin(orig_cursor.__class__, DbAliasLogger)
+            return PatchedCursor(orig_cursor.cursor, connection)
+
+        connection.cursor = cursor  # type: ignore[method-assign]
```
Comment (przlada) [inline saleor/core/db/connection.py:34]: I am not sure, but based on Django's code for `SimpleTestCase` class, we should also consider overriding these methods: `connect`, `temporary_connection,` and `chunked_cursor` but I could be wrong.

File: saleor/settings.py [hunk 3]
```diff
@@ -231,6 +240,20 @@ def get_url_from_env(name, *, schemes=None) -> Optional[str]:
     "saleor.core.middleware.jwt_refresh_token_middleware",
 ]
 
+# Database alias logger middleware
+ENABLE_DB_ALIAS_LOGGER_MIDDLEWARE = get_bool_from_env(
+    "ENABLE_DB_ALIAS_LOGGER_MIDDLEWARE", False
+)
+if ENABLE_DB_ALIAS_LOGGER_MIDDLEWARE:
+    MIDDLEWARE.append("saleor.core.db.connection.db_alias_logger_middleware")
```
Comment (przlada) [inline saleor/settings.py:248]:
```suggestion
    MIDDLEWARE = ["saleor.core.db.connection.db_alias_logger_middleware"] + MIDDLEWARE
```
Shouldn't we add this middleware as the first one to also wrap other middleware potentially?

**Comment by maarcingebala** [comment]
Closing in favor of https://github.com/saleor/saleor/pull/15428.

**Pull request closed without merging**